src/Cache.cpp

```diff
@@ -2945,6 +2946,26 @@ Cache::roomNamesAndAliases()
 {
     auto txn = ro_txn(env_);
 
+    auto getParentRoomIdsWithTxn = [&](const std::string &id) -> std::optional<std::string> {
+        auto cursor         = lmdb::cursor::open(txn, spacesParentsDb_);
+        std::string_view sp = id, space_parent;
+        if (cursor.get(sp, space_parent, MDB_SET)) {
+            while (cursor.get(sp, space_parent, MDB_FIRST_DUP)) {
+                if (!space_parent.empty())
+                    return std::make_optional(static_cast<std::string>(space_parent));
+            }
+        }
+        cursor.close();
+
+        return std::nullopt;
+    };
```

Comment:
This actually gets a random room parent. I would suggest you actually query the canonical room parent. Otherwise it might appear quite random, which room is actually listed here. I would suggest you check the parentSpace function in the TimelineModel.

Actually since you don't need the parentSpace for anything but displaying the name of the filtered entries, I would suggest you might want to do (in `RoomsModel::roleNames()`):

```cpp
auto parent = roomList->getRoomById();
if (parent) {
  auto desc = parent->parentSpace();
  if (desc) return desc->roomName();
}
```

Reply:
I'm not sure why, a room's parentSpace is empty. I tried to read off `mtx::events::state::space::Parent` events manually from a room and I did not find a parent. That's why I think I gave up and used spacesParentsDb.

This is the diff I used to test your approach

```cpp
case Roles::RoomParent: {
            const auto roomPtr = roomListModel_.getRoomById(QString::fromStdString(rooms[index.row()].id));
            if (auto &room = *roomPtr; roomPtr) {
                if (const auto &parent = room.parentSpace(); parent) {
                    qInfo() << "Parent has name" << parent->roomName();
                    return parent->roomName();
                } else {
                    qWarning() << "No parent for room" << "expected" << rooms[index.row()].parent;
                }
            }
            else {
                qWarning() << "No room with ID";
            }
            return QString{};
        }
```

Reply:
Did that room possibly not have a canonical parent? Can you maybe log the space parent event for that room?

Reply:
I don't think these rooms have any parents.

`cache::client()->getStateEventsWithType<mtx::events::state::space::Parent>(this->room_id_.toStdString()).size()` is 0.

Reply:
I think `spacesParentsDb_` builds this relation using `mtx::events::state::space::Child`. MSC2946 describes walking the spaces tree using `m.space.child`.

The spec also says

> Spaces form a hierarchy of rooms which clients can use to structure their room list into a tree-like view. The parent/child relationship can be defined in two ways: with [m.space.child](https://spec.matrix.org/v1.8/client-server-api/#mspacechild) state events in the space-room, or with [m.space.parent](https://spec.matrix.org/v1.8/client-server-api/#mspaceparent) state events in the child room.

> In most cases, both the child and parent relationship should be defined to aid discovery of the space and its rooms. When only a m.space.child is used, the space is effectively a curated list of rooms which the rooms themselves might not be aware of. When only a m.space.parent is used, the rooms are “secretly” added to spaces with the effect of not being advertised directly by the space.

> ...

> To avoid situations where a room falsely claims it is part of a given space, m.space.parent events should be ignored unless one of the following is true: A corresponding m.space.child event can be found in the supposed parent space. ...

I think m.space.child should be the primary sources of truth

Reply:
All 3 options are valid:

- only space.child
- only space.parent
- both

We generally only accept canonicalParents as something added to the room name in Nheko, because this means the room admin accepted that relationship. Otherwise a room can have multiple parents or someone could maliciously craft space.child events to make a room belong to a community it doesn't want to. So I think we really should only accept canonical parent events.

Reply:
In case the canonical parent events are not available for a room, can this be an opt-in config? This is the structure that [mautrix](https://github.com/mautrix) - matrix bridges use, so I'd really like to use the parent-child relationship from `spacesParentsDb_`. I think I can manually add the parent event to my ~50 matrix rooms but I'd prefer not to. 

Alternatively, I can make an issue on mautrix about this but I'm not aware of the technical decisions behind not doing this by default.

Reply:
I think we really should stick to only showing the canonical parent as the "official" parent. There are other features that rely on that like custom stickers and emojis. Maybe mautrix could change their spaces to use canonical parents?